### ISSUE TITLE
Support the new "Forwarded" header (RFC 7239)

### ISSUE THREAD

**Issue by cwe1ss** [aspnet/AspNetCore#5978]
There is now a standard for the most common X-Forwarded-\* headers. It was introduced with RFC 7239: "[Forwarded HTTP Extension](https://tools.ietf.org/html/rfc7239)". 

Example: `Forwarded: for=192.0.2.60;proto=http;by=203.0.113.43`

It would be great if this could be supported.

**Comment by colgreen** [comment]
> "Are you aware of any proxies that actually set that new header?"

Arguably that point is somewhat circular, i.e. maybe if ASP.NET Core supported the 'Forwarded' header there would be more pressure on the reverse proxy implementations to support it. It is after all the only approach that actually has a formal specification (rfc7239).

**Comment by damianh** [comment]
> Would also like to have ability to extend the 'Forwared' header with a `PathBase` parameter as well as support for that with the current implementation i.e. `X-Forwarded-PathBase`


Just to mention that I've shipped a package that supports this https://github.com/ProxyKit/HttpOverrides

**Comment by apeeters** [comment]
> > "Are you aware of any proxies that actually set that new header?"
> 
> Arguably that point is somewhat circular, i.e. maybe if ASP.NET Core supported the 'Forwarded' header there would be more pressure on the reverse proxy implementations to support it. It is after all the only approach that actually has a formal specification (rfc7239).

Indeed. If no one starts implementing the standard, it will never gain traction. Also: Spring has support for this `Forwarded` header since 2016...

**Comment by colgreen** [comment]
Perhaps someone could reach out to the YARP team to get their views, i.e. e.g. are they planning to implement rfc7239

**Comment by NatMarchand** [comment]
> Yes, we've seen it. Are you aware of any proxies that actually set that new header?

@Tratcher, it is the documented way for Azure APIM : https://docs.microsoft.com/en-us/azure/api-management/policies/set-header-to-enable-backend-to-construct-urls

**Comment by Tratcher** [comment]
> Perhaps someone could reach out to the YARP team to get their views, i.e. e.g. are they planning to implement rfc7239
> 
> [microsoft/reverse-proxy](https://github.com/microsoft/reverse-proxy)

YARP is being built by this team. We did add support for the new header in YARP but it's off by default. See https://microsoft.github.io/reverse-proxy/articles/transforms.html#forwarded

Design note: this probably should be implemented in AspNetCore as a whole new middleware, it takes very different options. I'm also not sure what to do with the randomized client ids, stash them in a new request feature?

**Comment by JonAnders** [comment]
> Design note: this probably should be implemented in AspNetCore as a whole new middleware, it takes very different options. 

@Tratcher I think there is value in keeping it in the same middleware, if possible, as the headers are conceptually the same. A class called `ForwardedHeadersMiddleware` is where I would expect to find the standards-compliant implementation, I wouldn't keep looking for a different middleware.

What if `Forwarded` was added to the `ForwardedHeaders` enum? To restrict the processing, in line with how `X-Forwarded-*` is configured, the enum could include `ForwardedBy`, `ForwardedFor`, `ForwardedHost` and `ForwardedProto` as well.

Support for both `Forwarded` and `X-Forwarded-*` could then be configured together, such as:
```
app.UseForwardedHeaders(new ForwardedHeadersOptions
{
    ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.ForwardedFor
});
```

There are some options that only make sense for the `X-Forwarded-*` headers, but is that a problem? They can either be ignored or cause an exception if they are set when only support for the `Forwarded` header is configured.


> I'm also not sure what to do with the randomized client ids, stash them in a new request feature?

Would it be a plausible alternative to just provide a class for parsing the `Forwarded` header and leave the rest up to the developer when using randomized client ids?

**Comment by Tratcher** [comment]
> Support for both `Forwarded` and `X-Forwarded-*` could then be configured together, such as:
> 
> ```
> app.UseForwardedHeaders(new ForwardedHeadersOptions
> {
>     ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.ForwardedFor
> });
> ```

That's a good example of why these shouldn't be combine into one middleware. If a request has both headers then what order should they be processed in? It's ambiguous.

**Comment by asbjornu** [comment]
> That's a good example of why these shouldn't be combine into one middleware. If a request has both headers then what order should they be processed in? It's ambiguous.

I'm not disagreeing with you @Tratcher, but I do think it's easy to argue that `Forwarded` should take precedence over `X-Forwarded-*`, both due to chronology and precision.

### Chronology

Since `X-Forwarded-*` existed before `Forwarded`, it is most likely that `Forwarded` was added after `X-Forwarded-*` in order to accommodate middleware that supports `Forwarded`. Middleware that doesn't support `Forwarded` will just ignore it and process `X-Forwarded-*` as before.

If adding `Forwarded` to an HTTP request doesn't do anything because `X-Forwarded-*` takes precedence, it makes it very hard to evolve beyond the status quo since every new header will have to be added as a clean break with existing infrastructure.

### Precision

`Forwarded` also has a precisely defined processing model, which I assume most implementers prefer over the undefined behavior of `X-Forwarded-*`.

What is a proxy supposed to do when there are an unequal amount of items added to `X-Forwarded-For`, `X-Forwarded-Host` and `X-Forwareded-Proto`, for instance? Which item corresponds to which between the three headers? Such situations may occur if a reverse proxy doesn't support all 3 headers and only modifies one or two of them.

There's also the question of whether new items should be appended or prepended to existing headers. There are numerous edge cases that are unresolved since these headers have been invented separately and never thoroughly specified in terms of a single, coherent standard.

Since `Forwarded` groups `for`, `by`, `host` and `proto` together for each item and clearly defines that items should be appended to an existing header, it doesn't suffer from these problems. If issues with `Forwarded` arises, IETF has a clearly defined process to improve upon the published RFC, which will be made available in a canonical RFC that all implementers can use and reference.

Because of all of these benefits, I think it's obvious that `Forwarded` should take precedence over `X-Forwarded-*`, when both exist in an HTTP request.

**Comment by Sebbs128** [comment]
I'd like to argue that neither should take precedence unless specifically configured to, but even then needing both should be rare. Whether `Forwarded` or `X-Forwarded-For` is the correct header to use depends very much upon the reverse proxy(s) in front of the web application. If the reverse proxy sends `Forwarded`, then configure the middleware to use `Forwarded` only. If they don't, then configure the middleware to use `X-Forwarded-For` (or `X-Real-IP`, `X-Azure-SocketIP`, or whatever the reverse proxy actually sends).

In a scenario where the reverse proxy appends `X-Forwarded-For` and not `Forwarded`, having `Forwarded` take precedence if both were (mis)configured (by a forward-thinking developer, or just plain misunderstanding) opens the door wide to spoofing via that header, with the real values being ignored/skipped.

**Comment by John0King** [comment]
it's about 10 yeas , any progress ?
